dll/cpl/desk/desktop.c

```diff
@@ -222,9 +333,86 @@ DesktopOnInitDialog(IN HWND hwndDlg, IN PDESKTOP_DATA pData)
         pData->bLocalHideChanged[i] = FALSE;
     }
 
+    pData->iLocalCurIcon = 0;
+    hwndList = GetDlgItem(hwndDlg, IDC_ICONS_LISTVIEW);
+    pData->hLocalImageList = ImageList_Create(GetSystemMetrics(SM_CXICON), GetSystemMetrics(SM_CYICON), ILC_COLOR32 | ILC_MASK, 1, 1);
+    ListView_SetImageList(hwndList, pData->hLocalImageList, LVSIL_NORMAL);
+
+    for (i = 0; i < _countof(IconChange); i++)
+    {
+        TCHAR szClassPath[MAX_PATH];
+        DWORD dwType, cbData;
+        LVITEM lvitem = {0};
+        HICON hIcon;
+
+        StringCchCopy(pData->LocalIcon[i].szPath, _countof(pData->LocalIcon[i].szPath), pData->Icon[i].szPath);
+        pData->bLocalIconChanged[i] = FALSE;
+
+        /* Try loading user-defined desktop icon title */
+        StringCchCopy(szClassPath, _countof(szClassPath), szUserClass);
+        StringCchCat(szClassPath, _countof(szClassPath), IconChange[i].CLSID);
+        cbData = sizeof(pData->LocalIcon[i].szTitle);
+
+        if (SHGetValue(HKEY_CURRENT_USER, szClassPath, IconChange[i].Icon, &dwType,
+                       pData->LocalIcon[i].szTitle, &cbData) != ERROR_SUCCESS || dwType != REG_SZ)
+        {
+            /* Try loading system-defined class icon title */
+            StringCchCopy(szClassPath, _countof(szClassPath), szSysClass);
+            StringCchCat(szClassPath, _countof(szClassPath), IconChange[i].CLSID);
+            cbData = sizeof(pData->LocalIcon[i].szTitle);
+
+            if (SHGetValue(HKEY_CLASSES_ROOT, szClassPath, IconChange[i].Icon, &dwType,
+                           pData->LocalIcon[i].szTitle, &cbData) != ERROR_SUCCESS || dwType != REG_SZ)
+            {
+                /* Fallback to predefined strings */
+                LoadString(GetModuleHandle(TEXT("shell32.dll")),
+                           IconChange[i].Title,
+                           pData->LocalIcon[i].szTitle,
+                           _countof(pData->LocalIcon[i].szTitle));
+            }
+        }
+
+        hIcon = GetIconFromLocation(pData->LocalIcon[i].szPath);
+
+        lvitem.mask = LVIF_TEXT | LVIF_IMAGE | LVIF_PARAM;
+        lvitem.iItem = i;
+        lvitem.iSubItem = 0;
+        lvitem.pszText = pData->LocalIcon[i].szTitle;
+        lvitem.lParam = (LPARAM)i;
+
+        if (hIcon)
+        {
+            if (pData->hLocalImageList)
+                lvitem.iImage = ImageList_AddIcon(pData->hLocalImageList, hIcon);
+            DestroyIcon(hIcon);
+        }
+
+        if (ListView_InsertItem(hwndList, &lvitem) >= 0)
+        {
+            if (i > 0)
```

Comment:
Why not make this test + continue outside of the `if ListView_InsertItem` ?

Reply:
We need to add all items to the list. Now I have restructured this a bit.